jb/itch5/array_based_order_book.hpp

```diff
@@ -0,0 +1,617 @@
+           << " px_begin_top_=" << px_begin_top_ << " px=" << px
+           << " qty=" << qty;
+        throw jb::feed_error(os.str());
+      }
+      // ... reduce the quantity ...
+      price_it->second -= qty;
+      if (price_it->second < 0) {
+        // ... this is "Not Good[tm]", somehow we missed an order or
+        // processed a delete twice ...
+        JB_LOG(warning) << "negative quantity in order book";
+      }
+      // now we can erase this element if updated qty <=0
+      if (price_it->second <= 0) {
+        bottom_levels_.erase(price_it);
+      }
+      return false;
+    }
+
+    // handles the top_levels_ price
+    if (better_(px, px_inside_)) {
+      std::ostringstream os;
+      os << "array_based_order_book::reduce_order."
+         << " Trying to reduce a non-existing top_levels_ price"
+         << " (better px_inside)."
+         << " px_begin_top_=" << px_begin_top_ << " px_inside_=" << px_inside_
+         << " px_end_top_=" << px_end_top_ << " px=" << px << " qty=" << qty;
+      throw jb::feed_error(os.str());
+    }
+    // get px relative position
+    auto rel_px = price_to_relative(px);
+    if (top_levels_.at(rel_px) == 0) {
+      std::ostringstream os;
+      os << "array_based_order_book::reduce_order."
+         << " Trying to reduce a non-existing top_levels_ price"
+         << " (top_levels_[rel_px] == 0)."
+         << " px_begin_top_=" << px_begin_top_ << " px_inside_=" << px_inside_
+         << " px_end_top_=" << px_end_top_ << " px relative position=" << rel_px
+         << " px=" << px << " qty=" << qty;
+      throw jb::feed_error(os.str());
+    }
+
+    // ... reduce the quantity ...
+    top_levels_.at(rel_px) -= qty;
+    if (top_levels_.at(rel_px) < 0) {
+      // ... this is "Not Good[tm]", somehow we missed an order or
+      // processed a delete twice ...
+      JB_LOG(warning) << "negative quantity in order book";
+      top_levels_.at(rel_px) = 0; // cant't be negative
+    }
+    // ... if it is not the inside we are done
+    if (px != px_inside_) {
+      return false;
+    }
+    // Is inside, ... now check if it was removed
+    if (top_levels_.at(rel_px) == 0) {
+      // gets the new inside (if any)
+      px_inside_ = next_best_price();
+      if (px_inside_ == empty_quote().first) {
+        // last top_levels_ price was removed...
+        // ... get the new inside from the bottom_levels
+        if (!bottom_levels_.empty()) {
+          auto new_inside_it = bottom_levels_.begin();
+          px_inside_ = new_inside_it->first;
+        }
+        // redefine limits
+        auto limits = get_limit_top_prices(px_inside_);
+        px_begin_top_ = std::get<0>(limits);
+        px_end_top_ = std::get<1>(limits);
+        // move tail prices from bottom_levels (px_begin_top_ {, .begin()})
+        move_bottom_to_top();
+      }
+    }
+    return true; // it is an inside change
+  }
+
+  /**
+   * Testing hook.
```

Comment:
ACK
Changed the name to price_comp.
I need the hook because the class is not a map, members (private) are.